Add parameter to ignore entities

q231950 opened the pull request:
Hi @rentzsch 

Recently I have been looking into using the _Xcode_ CoreData generated NSManagedObject subclasses. The code base we are working on is kind of big and we have **plenty** of _Mogenerator_ generated classes. 

In order to slowly migrate to the Xcode generated classes we need a way to use both generators at the same time. 

Xcode already supports this by allowing to enable/disable code generation per entity. This pull request adds support for ignoring entities when generating with _Mogenerator_ (I looked a bit but couldn't find any other solution to ignore entities).

Tests passed with `MacOSX10.14.sdk` and I added another test for checking if the parameter works.

Thank you and I am looking forward to your feedback!

rentzsch commented:
Hey Martin,

I like the exclusion idea for better Xcode interop and your code.

However, I think it would make more sense to opt out of mogenerator's codegen at the entity's userInfo level instead of passing in a new argument.

I'm thinking something like `mogenerator.skipgen` to match the `mogenerator.readonly` and `mogenerator.customBaseClass` naming scheme.

What do you think?

rentzsch commented:
Actually, I like your initial naming, so how about `mogenerator.ignore`?

Doesn't matter the value of the key, just the key's presence would be enough for mogenerator to ignore it.

q231950 commented:
I have added it to the entity now, it looks much better like this. I moved `isIgnored` into a new category: `@interface NSEntityDescription (userInfo)`. Please let me know if you're fine with that.

rentzsch commented:
Overall looks great.

I see two outstanding issues:

1) Most important: @atomicbird has been putting in a ton of work into the next release, and I'm fine with merging this current PR but want to clear with him that now's a good time to do it.

2) I don't see a reason now while there's two separate `:objc` tasks in test's `Rakefile`. Either I can just merge the `includes_uncle` lines into the first `:objc` task or you can do it. Unless I'm missing something.

q231950 commented:
1. No worries for getting this into this or another release. I'd be very happy when it gets in at all since I see that it really helps us and others might also benefit from it.

2. The idea was to have a specific test to assert that reading the entities' user info works. What do you think about having a separate method to have this assertion for both, the Swift and the Objective-C build?

```ruby
def assert_entity_user_info_respected
  # mogenerator.ignore
  includes_uncle = Dir.entries('./MOs').any? { |file| file.include?('Uncle') }
  raise 'Failed: Ignored entities should not be generated' if includes_uncle
end

desc 'Generate, Compile and Run Objective-C'
task :objc do
  Rake::Task[:clean].execute
  gen_and_compile_objc(MOGENERATOR_PATH, '', '')
  assert_entity_user_info_respected
  Rake::Task[:clean].execute
end

desc 'Generate, Compile and Run Swift'
task :swift do
  Rake::Task[:clean].execute
  gen_and_compile_swift(MOGENERATOR_PATH, '')
  assert_entity_user_info_respected
  Rake::Task[:clean].execute
end
```

q231950 commented:
@rentzsch, @atomicbird I updated all the things, looking forward to it, thank you 🙂

atomicbird commented:
I haven't had a chance to review this in detail but if it looks good to you @rentzsch then by all means merge it. Most of my recent work has gone into the `swift42` branch, and I'll deal with any conflicts that arise there.

q231950 commented:
Awesome @rentzsch @atomicbird, thanks for merging 🎉
